proxy/coco/src/request/waiting_room.rs

```diff
@@ -0,0 +1,700 @@
+    /// waiting room.
+    requests: HashMap<RadUrn, SomeRequest<T>>,
+
+    /// The configuration of the waiting room.
+    config: Config,
+}
+
+/// The `Conig` for the waiting room tells it what are the maximum number of query and clone
+/// attempts that can be made for a single request.
+///
+/// The recommended approach to initialising the `Config` is to use its `Default` implementation,
+/// i.e. `Config::default()`.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Config {
+    /// The maximum number of query attempts that can be made.
+    pub max_queries: Queries,
+    /// The maximum number of clone attempts that can be made.
+    pub max_clones: Clones,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            max_queries: MAX_QUERIES,
+            max_clones: MAX_CLONES,
+        }
+    }
+}
+
+/// The `Strategy` enumeration is for picking a strategy when selecting a `Request` from the
+/// [`WaitingRoom`]. Specifically, it's used in [`WaitingRoom::next`] and [`WaitingRoom::ready`]
+/// for selecting a `Request` that is in the state `IsCreated` and `Cloned`, respectively.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum Strategy<R> {
+    /// Select the first from the iterator.
+    First,
+    /// Select the newest from the iterator.
+    Newest,
+    /// Select the oldest from the iterator.
+    Oldest,
+    /// Select an item from the iterator at random.
+    Random(R),
+}
+
+impl<R> Strategy<R> {
+    /// Get back a `T` based off of the `Strategy`.
+    ///
+    /// If the iterator is empty then `None` is returned.
+    pub fn next<'a, T: 'a, U: Ord>(
+        self,
+        mut items: impl Iterator<Item = &'a T>,
+        mut key: impl FnMut(&T) -> U,
+    ) -> Option<&'a T>
+    where
+        R: Rng,
+    {
+        match self {
+            Self::First => items.next(),
+            Self::Newest => items.max_by_key(|i| key(i)),
+            Self::Oldest => items.min_by_key(|i| key(i)),
+            Self::Random(mut rng) => items.choose(&mut rng),
+        }
+    }
+}
+
+impl<T> WaitingRoom<T> {
+    /// Initialise a new `WaitingRoom` with the supplied `config`.
+    #[must_use]
+    pub fn new(config: Config) -> Self {
+        Self {
+            requests: HashMap::new(),
+            config,
+        }
+    }
+
+    /// Get the underlying [`SomeRequest`] for the given `urn.
+    ///
+    /// Returns `None` if there is no such request.
+    #[must_use]
+    pub fn get(&self, urn: &RadUrn) -> Option<&SomeRequest<T>> {
+        self.requests.get(urn)
+    }
+
+    /// Create a fresh [`Request`] with the given `urn`.
+    ///
+    /// If the `Request` already existed in the `WaitingRoom` then we get back the original request
+    /// as `Some`. This means we did nothing with the `urn` or `timestamp`.
+    pub fn create(&mut self, urn: RadUrn, timestamp: T) -> Option<SomeRequest<T>>
```

Comment:
```suggestion
    pub fn request(&mut self, urn: RadUrn, timestamp: T) -> Option<SomeRequest<T>>
```

Communicates more clearly: waiting_room request urn

Reply:
Well... Does it? It hasn't been _requested_ it's just created and is _ready_ to be requested.

Reply:
The caller doesn't know if it hasn't been created yet, the signature encodes that as well, returning `None` if the urn is already associated with an request. Which could be read as `Noop`.

Reply:
Ya, ok. It's a bit of grey area semantically for me because of the other state `IsRequested` but I can see your point. And since the next call on that request should be `queried` I'm happy to change this :+1:

Reply:
Ref. https://github.com/radicle-dev/radicle-upstream/pull/903#discussion_r493358673 to back the point.